### PR TITLE
listener: clean up accept filter before creating connection

### PR DESCRIPTION
**Pull request by lambdai**
Signed-off-by: Yuchen Dai silentdai@gmail.com

Description:
When listener continue to accept connection on listener filter timeout, the OS socket fd could be owned by two separate FileEventImpl.
The destructing FileEventImpl cannot recover the intention of the alive FileEventImpl.

The observation is that EdgeTrigger flag is cleared and the fd will be LevelTriggered and the EPOLLOUT will be activated in each epoll_wait
immediately after the invoke.

Envoy worker thread owning that connecting will be consuming 1 cpu core in the repeated epoll_wait.

This PR is maintaining the immutability that at most one FileEventImpl could owns the fd so the assigned event always match the expectation
of FileEventImpl.

Before this PR
```
 # registered by tls_inspector
1573028637.322859 epoll_ctl(14, EPOLL_CTL_ADD, 26, {EPOLLIN|EPOLLRDHUP|EPOLLET, {u32=26, u64=26}}) = 0 
# registered by HCM
1573028637.419820 epoll_ctl(14, EPOLL_CTL_MOD, 26, {EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, {u32=26, u64=26}}) = 0         
# tls_inspector clean up, EPOLLET is cleared somehow
1573028637.420488 epoll_ctl(14, EPOLL_CTL_MOD, 26, {EPOLLIN|EPOLLOUT, {u32=26, u64=26}}) = 0                                                   

#repeated immediate return of epoll_wait, no write since the http write buffer is empty at the beginning
1573028637.420537 epoll_wait(14, [{EPOLLOUT, {u32=26, u64=26}}], 32, 24) = 1
```

With this PR
```
# registered by tls_inspector
1573030976.461595 epoll_ctl(14, EPOLL_CTL_ADD, 26, {EPOLLIN|EPOLLRDHUP|EPOLLET, {u32=26, u64=26}}) = 0 
# tls_inspector clean up  
1573030976.562546 epoll_ctl(14, EPOLL_CTL_DEL, 26, 0x7f1d914cd0a0) = 0                                                                                          
# registered by HCM, EPOLLET show up and leave there forever
1573030976.562986 epoll_ctl(14, EPOLL_CTL_ADD, 26, {EPOLLIN|EPOLLOUT|EPOLLET, {u32=26, u64=26}}) = 0                       
```

The proof that `EPOLLIN|EPOLLOUT` w/o EPOLLET is registered due to ~FileEventImpl
Below `fd 21 op 3 event_uaddr 5` is recognized as `epoll_ctl( ..., EPOLL_MOD, 21, { EPOLLIN|EPOLLOUT, {...}}) ` where EPOLLET (1 << 31) is missing. 
fd 21 is the English spelling of 26 since I capture the logs in different run.

```
fd 21 op 3 event_uaddr 5 0x7f443e11d24a : epoll_ctl 0xa/0x30 [/lib/x86_64-linux-gnu/libc-2.27.so]
 0x562790a619b2 : epoll_apply_one_change 0x172/0x570 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x562790a61356 : epoll_nochangelist_del 0xa6/0xb0 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x562790a53b1c : evmap_io_del_ 0x34c/0x3e0 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x562790a48d15 : event_del_nolock_ 0x2d5/0x490 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x562790a4e60c : event_del_ 0xac/0x100 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x562790a4cc2a : event_del 0x1a/0x20 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56279038c145 : Envoy::Event::ImplBase::~ImplBase() 0x15/0x30 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x562790366047 : Envoy::Event::FileEventImpl::~FileEventImpl() 0x47/0x60 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56279036607c : Envoy::Event::FileEventImpl::~FileEventImpl() 0x1c/0x30 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56278f7b8a5f : std::default_delete<Envoy::Event::FileEvent>::operator()(Envoy::Event::FileEvent*) const 0x2f/0x40 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56278f7b83a3 : std::unique_ptr<Envoy::Event::FileEvent, std::default_delete<Envoy::Event::FileEvent> >::~unique_ptr() 0x53/0x80 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56278f7d0187 : Envoy::Extensions::ListenerFilters::TlsInspector::Filter::~Filter() 0x47/0x70 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56278f7d01cc : Envoy::Extensions::ListenerFilters::TlsInspector::Filter::~Filter() 0x1c/0x30 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56278f7b23bf : std::default_delete<Envoy::Network::ListenerFilter>::operator()(Envoy::Network::ListenerFilter*) const 0x2f/0x40 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56278f7b1cd3 : std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> >::~unique_ptr() 0x53/0x80 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x562790350879 : void __gnu_cxx::new_allocator<std::_List_node<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> > > >::destroy<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> > >(std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> >*) 0x19/0x20 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x5627903507d0 : void std::allocator_traits<std::allocator<std::_List_node<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> > > > >::destroy<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> > >(std::allocator<std::_List_node<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> > > >&, std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> >*) 0x20/0x30 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x56279035072e : std::__cxx11::_List_base<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> >, std::allocator<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> > > >::_M_clear() 0x6e/0xa0 [/home/lambdai/workspace/tls/envoymaster1106.dbg]
 0x5627903505cf : std::__cxx11::list<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> >, std::allocator<std::unique_ptr<Envoy::Network::ListenerFilter, std::default_delete<Envoy::Network::ListenerFilter> > > >::clear() 0x1f/0x40 [/home/lambdai/workspace/tls/envoymaster1106.dbg]

```

Risk Level:
Testing:
Docs Changes:
Release Notes:
Fix istio/istio#18229

### CODE REVIEW

File: source/server/connection_handler_impl.cc [hunk 1]
```diff
@@ -264,6 +264,9 @@ void ConnectionHandlerImpl::ActiveTcpSocket::newConnection() {
       socket_->setDetectedTransportProtocol(
           Extensions::TransportSockets::TransportProtocolNames::get().RawBuffer);
     }
+    // Erase accept filter states because accept filters may not get the opportunity to clean up.
```
Comment (mattklein123) [inline source/server/connection_handler_impl.cc]: Can you add a TODO here and file an issue around figuring out a more structural way of preventing this from happening? I think as we move towards IoHandles being produced by the dispatcher, it should be possible internally to figure out this case.

Comment (lambdai) [reply source/server/connection_handler_impl.cc]: will do

Comment (mattklein123) [reply source/server/connection_handler_impl.cc]: Is it possible to add a quick unit test that at least makes sure that an accept filter is destructed before the connection is created? I think that shouldn't be too hard? @lizan WDYT?

Comment (lambdai) [reply source/server/connection_handler_impl.cc]: It's super hard to link everything together.
I tried to to verify accept_filters empty but it requires real world internal structs ActiveTcpXXX
I also tried verify the sequence of operations of FileEvent but it also need some surgery.

@lizan  Any advice?

Comment (lambdai) [reply source/server/connection_handler_impl.cc]: I have a super simple solution. Hold on

File: test/mocks/network/mocks.cc [hunk 2]
```diff
@@ -97,8 +97,10 @@ MockUdpListenerCallbacks::~MockUdpListenerCallbacks() = default;
 MockDrainDecision::MockDrainDecision() = default;
 MockDrainDecision::~MockDrainDecision() = default;
 
-MockListenerFilter::MockListenerFilter() = default;
-MockListenerFilter::~MockListenerFilter() = default;
+MockListenerFilter::MockListenerFilter() {
+  // ON_CALL(*this, die_()).WillByDefault(testing::Invoke([](){}));
```
Comment (lambdai) [inline test/mocks/network/mocks.cc]: I don't really know why ON_CALL + NiceMock reports uninterested call.
I add explicit EXPECT_CALL in each usage. That's a tech debt.
I will investigate later

File: source/server/connection_handler_impl.cc (hunk 1 above)
Comment (lambdai) [reply source/server/connection_handler_impl.cc]: updated

File: source/server/connection_handler_impl.cc [hunk 3]
```diff
@@ -264,6 +264,11 @@ void ConnectionHandlerImpl::ActiveTcpSocket::newConnection() {
       socket_->setDetectedTransportProtocol(
           Extensions::TransportSockets::TransportProtocolNames::get().RawBuffer);
     }
+    // TODO(lambdai): add test cases
```
Comment (mattklein123) [inline source/server/connection_handler_impl.cc]: nit: update to add integration tests

File: test/mocks/network/mocks.cc (hunk 2 above)
Comment (mattklein123) [reply test/mocks/network/mocks.cc]: It doesn't work in the destructor of a mock. We do this all over the place. Please delete this comment.

File: test/mocks/network/mocks.h [hunk 4]
```diff
@@ -164,6 +164,7 @@ class MockListenerFilter : public ListenerFilter {
   MockListenerFilter();
   ~MockListenerFilter() override;
 
+  MOCK_METHOD0(die_, void());
```
Comment (mattklein123) [inline test/mocks/network/mocks.h]: s/die_/destroy which is what we typically name this method

File: test/server/connection_handler_test.cc [hunk 5]
```diff
@@ -361,7 +361,8 @@ TEST_F(ConnectionHandlerTest, NormalRedirect) {
   EXPECT_CALL(test_listener2->socket_, localAddress()).WillRepeatedly(ReturnRef(alt_address));
   handler_->addListener(*test_listener2);
 
-  Network::MockListenerFilter* test_filter = new Network::MockListenerFilter();
+  auto* test_filter = new NiceMock<Network::MockListenerFilter>();
+  EXPECT_CALL(*test_filter, die_()).Times(1);
```
Comment (mattklein123) [inline test/server/connection_handler_test.cc]: nit: remove Times(1) here and elsewhere which is redundant.

File: test/mocks/network/mocks.cc (hunk 2 above)
Comment (lambdai) [reply test/mocks/network/mocks.cc]: Sad...`During the constructor or destructor of MockFoo, the mock object is not nice or strict`

File: test/mocks/network/mocks.h (hunk 4 above)
Comment (lambdai) [reply test/mocks/network/mocks.h]: Updating...
Looks like `destroy_` w/ underscore is preferred since it is added in MockClass

**Comment by lambdai** [comment]
Thanks! Running test locally. Update in 5 minutes

**Pull request merged**